make/jdk/src/classes/build/tools/methodhandle/VarHandleGuardMethodGenerator.java

```diff
@@ -122,11 +122,14 @@ final class VarHandleGuards {
                         if (direct && handle.vform.methodType_table[ad.type] == ad.symbolicMethodTypeErased) {
                             <RESULT_ERASED>MethodHandle.linkToStatic(<LINK_TO_STATIC_ARGS>);<RETURN_ERASED>
                         } else {
-                            MethodHandle mh = handle.getMethodHandle(ad.mode);
-                            <RETURN>mh.asType(ad.symbolicMethodTypeInvoker).invokeBasic(<LINK_TO_INVOKER_ARGS>);
+                            <RETURN>ad.adaptedMethodHandle(handle).invokeBasic(<LINK_TO_INVOKER_ARGS>);
                         }
                     }""";
 
+    // Currently this void version is needed because otherwise
+    // TestZGCBarrierElision.testAtomicThenAtomicAnotherField fails
+    // However, testArrayAtomicThenAtomic, testAtomicThenAtomic, and
+    // testArrayAtomicThenAtomicAtUnknownIndices works
```

Comment:
It doesn't look right. Was is the root cause of the failure? Can it be a test bug?